solana/bridge-escrow/programs/bridge-escrow/src/lib.rs

```diff
@@ -0,0 +1,840 @@
+                ErrorCode::InvalidEventFormat
+            })?;
+        }
+
+        Ok(())
+    }
+}
+
+// Define the Auctioneer account
+#[account]
+pub struct Auctioneer {
+    pub authority: Pubkey,
+}
+
+// Define the Intent account with space calculation
+#[account]
+#[derive(Debug, PartialEq, Eq, InitSpace)]
+pub struct Intent {
+    #[max_len(20)]
+    pub intent_id: String,
+    // User on source chain
+    #[max_len(40)]
+    pub user_in: String,
+    // User on destination chain
+    pub user_out: Pubkey,
+    #[max_len(40)]
+    pub token_in: String,
+    pub amount_in: u64,
+    #[max_len(20)]
+    pub token_out: String,
+    #[max_len(20)]
+    pub amount_out: String,
+    pub winner_solver: Pubkey,
+    // Timestamp when the intent was created
+    pub creation_timestamp_in_sec: u64,
+    pub timeout_timestamp_in_sec: u64,
+    pub single_domain: bool,
+}
+
+#[derive(AnchorSerialize, AnchorDeserialize, Debug)]
+pub struct IntentPayload {
+    pub intent_id: String,
+    pub user_in: String,
+    pub user_out: Pubkey,
+    pub token_in: String,
+    pub amount_in: u64,
+    pub token_out: String,
+    pub amount_out: String,
+    pub winner_solver: Pubkey,
+    pub timeout_timestamp_in_sec: u64,
+    pub single_domain: bool,
+}
+
+// Define the context for initializing the program
+#[derive(Accounts)]
+#[instruction()]
+pub struct Initialize<'info> {
+    #[account(mut)]
+    pub authority: Signer<'info>,
+    #[account(init, seeds = [AUCTIONEER_SEED], bump, payer = authority, space = 8 + 32)]
+    pub auctioneer: Account<'info, Auctioneer>,
+
+    #[account(init, payer = authority, seeds = [DUMMY_SEED], bump, mint::decimals = 9, mint::authority = auctioneer)]
+    pub token_mint: Account<'info, Mint>,
+
+    pub system_program: Program<'info, System>,
+    pub token_program: Program<'info, Token>,
+    pub associated_token_program: Program<'info, AssociatedToken>,
+}
+
+#[derive(Accounts)]
+#[instruction(intent_id: String)]
+pub struct ReceiveTransferContext<'info> {
+    #[account(mut)]
+    pub authority: Signer<'info>,
+
+    #[account(seeds = [AUCTIONEER_SEED], bump, has_one = authority)]
+    pub auctioneer_state: Account<'info, Auctioneer>,
+
+    pub token_mint: Account<'info, Mint>,
+    #[account(mut, token::mint = token_mint, token::authority = auctioneer_state)]
+    pub escrow_token_account: Account<'info, TokenAccount>,
+    #[account(mut, token::mint = token_mint)]
+    pub solver_token_account: Account<'info, TokenAccount>,
+
+    #[account(address = solana_program::sysvar::instructions::ID)]
+    /// CHECK: Used for getting the caller program id to verify if the right
+    /// program is calling the method.
+    pub instruction: UncheckedAccount<'info>,
+
+    pub token_program: Program<'info, Token>,
+
+    // New Intent account addition
+    #[account(
+        mut, 
+        seeds = [b"intent", intent_id.as_bytes()], 
+        bump,
+        constraint = intent.user_in == authority.key().to_string()
+    )]
```

Comment:
The `user_in` is the user who escrowed the funds but the caller is the solana-ibc contract or auctioneer.
```suggestion
    #[account(
        mut, 
    )]
```